src/Pos/DB/Block.hs

```diff
@@ -102,6 +103,31 @@ loadDataWhile getter predicate start = NewestFirst <$> doIt start
                 then (d :) <$> doIt prev
                 else pure []
 
+-- Return all blocks that match the given predicate. It is repetetive,
+-- but it's also already pretty generic.
+loadDataMatching
+    :: forall m a .
+       (Monad m, HasPrevBlock a)
+    => (HeaderHash -> m a)
+    -> (a -> Bool)
+    -> HeaderHash
+    -> m (NewestFirst [] a)
+loadDataMatching getter predicate start = NewestFirst <$> doIt start []
+  where
+    doIt :: HeaderHash -> [HeaderHash] -> m [a]
+    doIt h acc
+        -- When we reach the genesis block, return the accumulator.
+        | h == genesisHash = sequence $ getter <$> acc
```

Comment:
It's inefficient to use `getter` twice, because it accesses DB and it can be slow. Why not store `[a]` in accumulator?